rust/operator-binary/src/main.rs

```diff
@@ -0,0 +1,128 @@
+    cmd: Cmd,
+}
+
+#[derive(StructOpt)]
+enum Cmd {
+    /// Print CRD objects
+    Crd,
+    /// Run operator
+    Run {
+        /// Provides the path to a product-config file
+        #[structopt(long, short = "p", value_name = "FILE")]
+        product_config: Option<String>,
+    },
+}
+
+mod built_info {
+    // The file has been placed there by the build script.
+    include!(concat!(env!("OUT_DIR"), "/built.rs"));
+}
+
+/// Erases the concrete types of the controller result, so that we can merge the streams of multiple controllers for different resources.
+///
+/// In particular, we convert `ObjectRef<K>` into `ObjectRef<DynamicObject>` (which carries `K`'s metadata at runtime instead), and
+/// `E` into the trait object `anyhow::Error`.
+fn erase_controller_result_type<K: Resource, E: std::error::Error + Send + Sync + 'static>(
+    res: Result<(ObjectRef<K>, ReconcilerAction), E>,
+) -> anyhow::Result<(ObjectRef<DynamicObject>, ReconcilerAction)> {
+    let (obj_ref, action) = res?;
+    Ok((obj_ref.erase(), action))
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    stackable_operator::logging::initialize_logging("SPARK_OPERATOR_LOG");
+    let opts = Opts::from_args();
+    match opts.cmd {
+        Cmd::Crd => println!("{}", serde_yaml::to_string(&SparkCluster::crd())?,),
+        Cmd::Run { product_config } => {
+            stackable_operator::utils::print_startup_string(
+                built_info::PKG_DESCRIPTION,
+                built_info::PKG_VERSION,
+                built_info::GIT_VERSION,
+                built_info::TARGET,
+                built_info::BUILT_TIME_UTC,
+                built_info::RUSTC_VERSION,
+            );
+            let product_config = if let Some(product_config_path) = product_config {
+                ProductConfigManager::from_yaml_file(&product_config_path)?
+            } else {
+                ProductConfigManager::from_str(include_str!(
+                    "../../../deploy/config-spec/properties.yaml"
+                ))?
+            };
+            let client =
+                stackable_operator::client::create_client(Some("spark.stackable.tech".to_string()))
+                    .await?;
+            let controller_builder =
+                Controller::new(client.get_all_api::<SparkCluster>(), ListParams::default());
+            let sc_store = controller_builder.store();
+            let controller = controller_builder
+                .owns(client.get_all_api::<Service>(), ListParams::default())
+                .watches(
+                    client.get_all_api::<Endpoints>(),
+                    ListParams::default(),
+                    move |endpoints| {
+                        sc_store
+                            .state()
+                            .into_iter()
+                            .filter(move |sc| {
+                                let _ = &endpoints; // capture endpoints to prevent it from being dropped (2021 warning)
```

Comment:
It happens when building with `-W rust-2021-compatibility` which has been added to the GH Rust action recently.

Reply:
To be clear: this warning requires the explicit `-D rust-2021-compatibility` flag. However, it's also a false positive since there are no side effects to dropping an `Endpoints` (aside from reclaiming some memory).